### ISSUE TITLE
Add actor reminder support for Cron expressions

### ISSUE THREAD

**Issue by WhitWaldo** [dapr/dapr#7894]
## In what area(s)?
/area runtime

## Describe the feature
Today, actor reminders are scheduled using a timespan; that is, how long from now should the reminder be triggered. This is generally fine when something needs to happen every 15 minutes or some other regular interval, but it's less ideal for date-specific reminders (e.g. this needs to be triggered on the second day of every month at some specified time).  Today, I handle this by picking a time on the target day, working out the time span interval between then and now and then register a non-recurring reminder to trigger at that moment.

There are two key problems with this approach:
1) If there's a system fault of any sort between when the reminder is triggered and when it re-registers the next non-recurring reminder, future reminders will not execute at all. Today, I handle this by having _another_ reminder check every week on a recurring schedule to verify that each of the monthly reminders that _should_ be set, actually are (so as to avoid the re-registration problem), but this puts preventable load on the actor and scheduling system.
2) I have to store additional state into the reminder to ensure that I'm always firing it at the same time each day (I usually pick random hour/minute combinations to avoid having _all_ my actors execute at the same time if the operation is just date-specific).

As the distributed scheduler is being released in the near future and because it already is tightly coupled with workflows (and thus actors), I'd like to request an overload to reminders that accepts a Cron expression which should be compatible with the distributed scheduler. This would enable a situation where I can set a random time on the intended date only once and should something happen that prevents the process from completing successfully on the reminder invocation, it has a standing registration, so it'll just fire again at the next interval without the verification check.

From a .NET perspective, I envision the overloads would look awfully similar to what's already configured. We already have the following:

```cs
RegisterReminderAsync(string reminderName, byte[] state, TimeSpan dueTime, TimeSpan period);

RegisterReminderAsync(string reminderName, byte[] state, TimeSpan dueTime, TimeSpan period, int repetitions);

RegisterReminderAsync(string reminderName, byte[] state, TimeSpan dueTime, TimeSpan period, int repititions, TimeSpan ttl);
```

I would propose the following as well (note I remove the dueTime and period as a Cron expression is really like both in one saying this is when it should execute ):
```cs
// Note I only have a period and not a dueTime - this is because a Cron expression is really like both saying execute every time when the current date/time matches this constraint
RegisterReminderAsync(string reminderName, byte[] state, string periodCron);

//This differs from the previous overload in that it limits the operation to a fixed number of repetitions
RegisterReminderAsync(string reminderName, byte[] state, string periodCron, int repetitions);
```

I can't think of a situation where you'd need a TTL if you have a Cron-style schedule and a fixed repetition count (meaning, you already easily know when it'll be expired), so I left off that overload.

## Release Note
RELEASE NOTE: **ADD** Can schedule actor reminders using Cron expressions

**Comment by olitomlinson** [comment]
+1 for this - this is a significant Developer Experience improvement which can be brought to all languages that support Actors!

The overload you had in mind is not dissimilar to what I was thinking too :)

**Comment by yaron2** [comment]
Good idea, but since Distributed Scheduler also brings a new API that is completely decoupled from actor reminders and workflows, it might be good to add the cron support there to start with. You can see the jobs API here: https://github.com/dapr/docs/blob/v1.14/daprdocs/content/en/reference/api/jobs_api.md

Essentially we would add a cron expression field to duetime or replace it with it.

**Comment by WhitWaldo** [comment]
@yaron2 I'm going to have to see how that documentation evolves, because I thoroughly don't follow what precisely a job _is_ from it.

**Comment by yaron2** [comment]
> @yaron2 I'm going to have to see how that documentation evolves, because I thoroughly don't follow what precisely a job _is_ from it.

https://v1-14.docs.dapr.io/developing-applications/building-blocks/jobs/jobs-overview/

**Comment by WhitWaldo** [comment]
@yaron2 I've been thinking about this and I think I'd still like to have this added to the actors themselves instead of just using a job. The use case I have in mind is an actor that has to perform an action every month on the second day. Sure, I could write a controller endpoint on the type and invoke it via a job, call into the actor and have it do whatever, but it now requires that I have code in two places instead of the one actor class.

In the interest of keeping actor code separate from that of the rest of the project, I'd still like to see this be an actor-specific add-in, even if that's precisely how it works under the hood.

**Comment by olitomlinson** [comment]
> I'd still like to see this be an actor-specific overload, even if under the hood, it's directly invoking the actor method via an HTTP invocation.

I agree. The best DevEx is to use an overload, rather than making the Dev care about another thing which is outside the scope of the Actor Framework.

This may cause friction for a little while if users don't realise that this overload would only work if the distributed scheduler is available/enabled. But that can be addressed with suitable documentation.

**Comment by yaron2** [comment]
Since the scheduler service is the subsystem for both reminders and the jobs API it would be very easy to add cron support for both and keep the API consistent
